### PR TITLE
[components] Add --rebuild-component-registry option, `--skip-venv` option, build cache on code loc construction

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Add a top-level `--rebuild-component-registry` option that clears the cache and refetches the component registry for the current environment. The reason this is called `--rebuild-component-registry` instead of `--rebuild-cache` is:

- It is scoped to just the current environment, whereas the cache encompasses multiple environments
- We may cache other info later

Also add `--skip-venv` option to `dg generate code-location` to avoid automatic venv setup and cache population on construction. Mostly useful for testing.

We also invoke this immediately after `uv sync` when constructing a new code location, so that the cache is loaded.

## How I Tested These Changes

New unit tests + manual test of code location generation.

### CODE REVIEW

**Comment by smackesey** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26591?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26591** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26591?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26591?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26478** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26478?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26590** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26590?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/libraries/dagster-dg/dagster_dg/cache.py [hunk 1]
```diff
@@ -47,9 +47,15 @@ def __init__(self, root_path: Path, logging_enabled: bool):
...
-        self.log(f"CACHE [clear]: {self._root_path}")
+    def clear(self, key: Optional[Tuple[str, ...]] = None) -> None:
+        if key:
+            path = self._get_path(key)
+            if path.exists():
+                path.unlink()
+                self.log(f"CACHE [clear]: {path}")
+        else:
+            shutil.rmtree(self._root_path)
+            self.log(f"CACHE [clear]: {self._root_path}")
```
Comment (schrockn) [inline python_modules/libraries/dagster-dg/dagster_dg/cache.py:58]: I think we should have two methods here. One "clear all" and one is "clear for specific key"

Super dangerous in general to have this structutre where if you forget to add a param it does a big destructive operation.

**Review by schrockn — CHANGES_REQUESTED** [review]
Make it harder to delete all the things

**Comment by smackesey** [comment]
I changed the key-specific clear to a new method called `delete`

File: python_modules/libraries/dagster-dg/dagster_dg/cache.py [hunk 2]
```diff
@@ -47,6 +47,12 @@ def __init__(self, root_path: Path, logging_enabled: bool):
         self._root_path.mkdir(parents=True, exist_ok=True)
         self._logging_enabled = logging_enabled
 
+    def delete(self, key: Tuple[str, ...]) -> None:
```
Comment (schrockn) [inline python_modules/libraries/dagster-dg/dagster_dg/cache.py:50]: clear and delete are synonyms.

how about `clear_key` and `clear_all` or something

**Review by schrockn — CHANGES_REQUESTED** [review]
one more naming round

File: python_modules/libraries/dagster-dg/dagster_dg/cache.py (hunk 2 above)
Comment (smackesey) [reply python_modules/libraries/dagster-dg/dagster_dg/cache.py]: changed to `clear_key` and `clear_all`

File: python_modules/libraries/dagster-dg/dagster_dg/cli/generate.py [hunk 3]
```diff
@@ -65,9 +65,15 @@ def generate_deployment_command(path: Path) -> None:
         " the location of the local Dagster clone will be read from the `DAGSTER_GIT_REPO_DIR` environment variable."
     ),
 )
+@click.option(
+    "--skip-venv",
+    is_flag=True,
+    default=False,
+    help="Do not create a virtual environment for the code location.",
+)
```
Comment (schrockn) [inline python_modules/libraries/dagster-dg/dagster_dg/cli/generate.py:73]: excellent

**Review by schrockn — APPROVED** [review]

**Pull request merged**